File: bench/coquic-python-perf/coquic_python_perf/io.py
from __future__ import annotations

import asyncio
import errno
import ipaddress
import socket
import sys
import time
from dataclasses import dataclass

import coquic

from . import PerfError

MAX_UDP_DATAGRAM_SIZE = 64 * 1024
MAX_BUFFERED_SEND_DATAGRAMS = 4096
_LINUX_IP_MTU_DISCOVER = 10
_LINUX_IP_PMTUDISC_PROBE = 3
_LINUX_IPV6_MTU_DISCOVER = 23


@dataclass(slots=True)
class Route:
    peer: tuple[str, int]
    address_validation_identity: bytes


@dataclass(slots=True)
class TxDatagram:
    route_handle: int
    bytes: bytes
    ecn: coquic.EcnCodepoint
    is_pmtu_probe: bool


@dataclass(slots=True)
class RxDatagram:
    bytes: bytes
    route_handle: int
    address_validation_identity: bytes


class _DatagramProtocol(asyncio.DatagramProtocol):
    def __init__(self) -> None:
        """Create a datagram protocol with an in-memory receive queue."""
        self.queue: asyncio.Queue[tuple[bytes, tuple[str, int]]] = asyncio.Queue()

    def datagram_received(self, data: bytes, addr) -> None:
        self.queue.put_nowait((bytes(data), addr))


class WaitEvent:
    def __init__(self, kind: str, datagram: RxDatagram | None = None):
        """Create a wait result for received datagrams, timers, or idle timeouts."""
        self.kind = kind
        self.datagram = datagram


class UdpRuntime:
    def __init__(
        self, transport: asyncio.DatagramTransport, protocol: _DatagramProtocol
    ):
        """Wrap an asyncio UDP endpoint with CoQUIC route bookkeeping."""
        self.transport = transport
        self.protocol = protocol
        self.start = time.monotonic()
        self.routes_by_handle: dict[int, Route] = {}
        self.handles_by_peer: dict[tuple[str, int], int] = {}
        self.next_route_handle = 1
        self.send_buffer: list[TxDatagram] = []

    @classmethod
    async def client(cls, host: str, port: int) -> tuple["UdpRuntime", int, bytes]:
        peer = _resolve_remote(host, port)
        bind_addr = _client_bind_address(peer)
        sock = _open_udp_socket(bind_addr, 0)
        loop = asyncio.get_running_loop()
        transport, protocol = await loop.create_datagram_endpoint(
            _DatagramProtocol,
            sock=sock,
        )
        runtime = cls(transport, protocol)
        route = runtime.ensure_route(peer)
        identity = runtime.address_validation_identity(route)
        if identity is None:
            raise PerfError("missing client primary route identity")
        return runtime, route, identity

    @classmethod
    async def server(cls, host: str, port: int) -> "UdpRuntime":
        loop = asyncio.get_running_loop()
        sock = _open_udp_socket(host, port)
        transport, protocol = await loop.create_datagram_endpoint(
            _DatagramProtocol,
            sock=sock,
        )
        return cls(transport, protocol)

    def close(self) -> None:
        self.transport.close()

    def now_us(self) -> int:
        return min(int((time.monotonic() - self.start) * 1_000_000), (1 << 64) - 1)

    def ensure_route(self, peer: tuple[str, int]) -> int:
        peer = (peer[0], int(peer[1]))
        if peer in self.handles_by_peer:
            return self.handles_by_peer[peer]
        handle = self.next_route_handle
        self.next_route_handle += 1
        self.handles_by_peer[peer] = handle
        self.routes_by_handle[handle] = Route(peer, address_validation_identity(peer))
        return handle

    def inbound_datagram(self, rx: RxDatagram) -> coquic.InboundDatagram:
        return coquic.InboundDatagram(
            bytes=rx.bytes,
            route_handle=rx.route_handle,
            address_validation_identity=rx.address_validation_identity,
            ecn=coquic.EcnCodepoint.UNAVAILABLE,
        )

    def append_result_sends(self, result: coquic.QueryResult) -> None:
        for effect in result.effects:
            if effect.kind != "send_datagram":
                continue
            if len(self.send_buffer) >= MAX_BUFFERED_SEND_DATAGRAMS:
                raise PerfError(
                    "send buffer exceeded before flush; call flush_sends more often"
                )
            if effect.route_handle is None:
                raise PerfError("send datagram missing route handle")
            self.send_buffer.append(
                TxDatagram(
                    route_handle=effect.route_handle,
                    bytes=effect.bytes,
                    ecn=effect.ecn,
                    is_pmtu_probe=effect.is_pmtu_probe,
                )
            )

    async def flush_sends(self) -> None:
        datagrams = self.send_buffer
        self.send_buffer = []
        for datagram in datagrams:
            route = self.routes_by_handle.get(datagram.route_handle)
            if route is None:
                raise PerfError(f"unknown route handle {datagram.route_handle}")
            self.transport.sendto(datagram.bytes, route.peer)
        await asyncio.sleep(0)

    async def recv(self) -> RxDatagram:
        data, peer = await self.protocol.queue.get()
        if len(data) > MAX_UDP_DATAGRAM_SIZE:
            data = data[:MAX_UDP_DATAGRAM_SIZE]
        route_handle = self.ensure_route(peer)
        identity = self.address_validation_identity(route_handle) or b""
        return RxDatagram(data, route_handle, identity)

    async def wait(self, next_wakeup: int | None, idle_timeout: float) -> WaitEvent:
        timer_timeout: float | None = None
        if next_wakeup is not None:
            now = self.now_us()
            if next_wakeup <= now:
                return WaitEvent("timer")
            timer_timeout = (next_wakeup - now) / 1_000_000.0
        timeout = timer_timeout if timer_timeout is not None else idle_timeout
        try:
            datagram = await asyncio.wait_for(self.recv(), timeout)
            return WaitEvent("datagram", datagram)
        except asyncio.TimeoutError:
            if timer_timeout is not None:
                return WaitEvent("timer")
            return WaitEvent("idle")

    def address_validation_identity(self, route_handle: int) -> bytes | None:
        route = self.routes_by_handle.get(route_handle)
        return route.address_validation_identity if route is not None else None


def copy_non_send_effects(result: coquic.QueryResult) -> list[coquic.Effect]:
    out = []
    for effect in result.effects:
        if effect.kind in (
            "receive_stream_data",
            "state_event",
            "connection_lifecycle_event",
            "peer_reset_stream",
            "peer_stop_sending",
        ):
            out.append(effect)
    return out


def address_validation_identity(peer: tuple[str, int]) -> bytes:
    address = ipaddress.ip_address(peer[0])
    port = int(peer[1])
    if isinstance(address, ipaddress.IPv4Address):
        return b"\x04" + address.packed + port.to_bytes(2, "big")
    return b"\x06" + address.packed + port.to_bytes(2, "big")


def _client_bind_address(peer: tuple[str, int]) -> str:
    address = ipaddress.ip_address(peer[0])
    if address.is_loopback:
        return "127.0.0.1" if isinstance(address, ipaddress.IPv4Address) else "::1"
    family = (
        socket.AF_INET
        if isinstance(address, ipaddress.IPv4Address)
        else socket.AF_INET6
    )
    with socket.socket(family, socket.SOCK_DGRAM) as probe:
        probe.connect(peer)
        local_host = probe.getsockname()[0]
    return str(ipaddress.ip_address(local_host))


def _resolve_remote(host: str, port: int) -> tuple[str, int]:
    infos = socket.getaddrinfo(host, port, type=socket.SOCK_DGRAM)
    if not infos:
        raise PerfError("failed to resolve remote address")
    addr = infos[0][4]
    return (addr[0], int(addr[1]))


def _open_udp_socket(host: str, port: int) -> socket.socket:
    infos = socket.getaddrinfo(host, port, type=socket.SOCK_DGRAM)
    if not infos:
        raise PerfError("failed to resolve bind address")

    family, _, _, _, sockaddr = infos[0]
    sock = socket.socket(family, socket.SOCK_DGRAM)
    try:
        if family == socket.AF_INET6 and hasattr(socket, "IPV6_V6ONLY"):
            sock.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_V6ONLY, 0)
        _configure_no_ip_fragmentation(sock)
        sock.bind(sockaddr)
        sock.setblocking(False)
        return sock
    except Exception:
        sock.close()
        raise


def _configure_no_ip_fragmentation(sock: socket.socket) -> None:
    ip_mtu_discover = getattr(socket, "IP_MTU_DISCOVER", None)
    ip_pmtudisc_probe = getattr(socket, "IP_PMTUDISC_PROBE", None)
    ipv6_mtu_discover = getattr(socket, "IPV6_MTU_DISCOVER", None)
    ipv6_pmtudisc_probe = getattr(socket, "IPV6_PMTUDISC_PROBE", None)

    if sys.platform.startswith("linux"):
        ip_mtu_discover = ip_mtu_discover or _LINUX_IP_MTU_DISCOVER
        ip_pmtudisc_probe = ip_pmtudisc_probe or _LINUX_IP_PMTUDISC_PROBE
        ipv6_mtu_discover = ipv6_mtu_discover or _LINUX_IPV6_MTU_DISCOVER
        ipv6_pmtudisc_probe = ipv6_pmtudisc_probe or _LINUX_IP_PMTUDISC_PROBE

    if ip_mtu_discover is not None and ip_pmtudisc_probe is not None:
        _set_socket_option_if_available(sock, socket.IPPROTO_IP, ip_mtu_discover, ip_pmtudisc_probe)
    if ipv6_mtu_discover is not None and ipv6_pmtudisc_probe is not None:
        _set_socket_option_if_available(
            sock, socket.IPPROTO_IPV6, ipv6_mtu_discover, ipv6_pmtudisc_probe
        )


def _set_socket_option_if_available(
    sock: socket.socket, level: int, option: int, value: int
) -> None:
    try:
        sock.setsockopt(level, option, value)
    except OSError as error:
        if error.errno not in (errno.ENOPROTOOPT, errno.EINVAL):
            raise

File: bench/coquic-go-perf/internal/io/io.go
package perfio

import (
	"fmt"
	"net"
	"strconv"
	"time"

	coquic "github.com/minhuw/coquic/bindings/go/coquic"
)

const (
	MaxUDPDatagramSize       = 64 * 1024
	MaxBufferedSendDatagrams = 4096
)

type Route struct {
	Peer                      *net.UDPAddr
	AddressValidationIdentity []byte
}

type TxDatagram struct {
	RouteHandle coquic.RouteHandle
	Bytes       []byte
	Ecn         coquic.EcnCodepoint
	IsPMTUProbe bool
}

type RxDatagram struct {
	Bytes                     []byte
	RouteHandle               coquic.RouteHandle
	AddressValidationIdentity []byte
}

type UdpRuntime struct {
	socket          *net.UDPConn
	start           time.Time
	routesByHandle  map[coquic.RouteHandle]Route
	handlesByPeer   map[string]coquic.RouteHandle
	nextRouteHandle coquic.RouteHandle
	sendBuffer      []TxDatagram
	recvBufferSize  int
}

type WaitKind int

const (
	WaitDatagram WaitKind = iota
	WaitTimer
	WaitIdle
)

type WaitEvent struct {
	Kind     WaitKind
	Datagram RxDatagram
}

func NewClient(host string, port uint16, recvBufferSize uint64) (*UdpRuntime, coquic.RouteHandle, []byte, error) {
	peer, err := resolveRemote(host, port)
	if err != nil {
		return nil, 0, nil, err
	}
	bindAddr := &net.UDPAddr{IP: net.IPv4zero, Port: 0}
	if peer.IP.To4() == nil {
		bindAddr = &net.UDPAddr{IP: net.IPv6zero, Port: 0}
	}
	socket, err := net.ListenUDP("udp", bindAddr)
	if err != nil {
		return nil, 0, nil, err
	}
	if err := configureNoIPFragmentation(socket); err != nil {
		socket.Close()
		return nil, 0, nil, err
	}
	runtime := newRuntime(socket, recvBufferSize)
	route := runtime.EnsureRoute(peer)
	identity := runtime.AddressValidationIdentity(route)
	if identity == nil {
		socket.Close()
		return nil, 0, nil, fmt.Errorf("missing client primary route identity")
	}
	return runtime, route, append([]byte(nil), identity...), nil
}

func NewServer(host string, port uint16, recvBufferSize uint64) (*UdpRuntime, error) {
	addr, err := net.ResolveUDPAddr("udp", net.JoinHostPort(host, strconv.Itoa(int(port))))
	if err != nil {
		return nil, err
	}
	socket, err := net.ListenUDP("udp", addr)
	if err != nil {
		return nil, err
	}
	if err := configureNoIPFragmentation(socket); err != nil {
		socket.Close()
		return nil, err
	}
	return newRuntime(socket, recvBufferSize), nil
}

func (r *UdpRuntime) Close() error {
	if r == nil || r.socket == nil {
		return nil
	}
	return r.socket.Close()
}

func (r *UdpRuntime) NowUs() coquic.TimeUs {
	return coquic.TimeUs(time.Since(r.start).Microseconds())
}

func (r *UdpRuntime) EnsureRoute(peer *net.UDPAddr) coquic.RouteHandle {
	key := peer.String()
	if handle, ok := r.handlesByPeer[key]; ok {
		return handle
	}
	handle := r.nextRouteHandle
	r.nextRouteHandle++
	r.handlesByPeer[key] = handle
	r.routesByHandle[handle] = Route{
		Peer:                      peer,
		AddressValidationIdentity: addressValidationIdentity(peer),
	}
	return handle
}

func (r *UdpRuntime) InboundDatagram(rx RxDatagram) coquic.InboundDatagram {
	return coquic.InboundDatagram{
		Bytes:                     rx.Bytes,
		RouteHandle:               rx.RouteHandle,
		HasRouteHandle:            true,
		AddressValidationIdentity: rx.AddressValidationIdentity,
		Ecn:                       coquic.EcnUnavailable,
	}
}

func (r *UdpRuntime) CollectResultEffects(result *coquic.QueryResult) ([]coquic.Effect, error) {
	out := make([]coquic.Effect, 0)
	err := result.ForEachEffect(func(effect coquic.Effect) error {
		switch effect.Kind {
		case coquic.EffectSendDatagram:
			if len(r.sendBuffer) >= MaxBufferedSendDatagrams {
				return fmt.Errorf("send buffer exceeded before flush; call flush_sends more often")
			}
			if !effect.HasRouteHandle {
				return fmt.Errorf("send datagram missing route handle")
			}
			r.sendBuffer = append(r.sendBuffer, TxDatagram{
				RouteHandle: effect.RouteHandle,
				Bytes:       effect.Bytes,
				Ecn:         effect.Ecn,
				IsPMTUProbe: effect.IsPMTUProbe,
			})
		case coquic.EffectReceiveStreamData,
			coquic.EffectStateEvent,
			coquic.EffectConnectionLifecycleEvent,
			coquic.EffectPeerResetStream,
			coquic.EffectPeerStopSending:
			out = append(out, effect)
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

func (r *UdpRuntime) FlushSends() error {
	datagrams := r.sendBuffer
	r.sendBuffer = nil
	for _, datagram := range datagrams {
		route, ok := r.routesByHandle[datagram.RouteHandle]
		if !ok {
			return fmt.Errorf("unknown route handle %d", datagram.RouteHandle)
		}
		_, err := r.socket.WriteToUDP(datagram.Bytes, route.Peer)
		if err != nil {
			return err
		}
	}
	return nil
}

func (r *UdpRuntime) SendBufferEmpty() bool {
	return len(r.sendBuffer) == 0
}

func (r *UdpRuntime) Recv() (RxDatagram, error) {
	buffer := make([]byte, r.recvBufferSize)
	length, peer, err := r.socket.ReadFromUDP(buffer)
	if err != nil {
		return RxDatagram{}, err
	}
	buffer = buffer[:length]
	route := r.EnsureRoute(peer)
	identity := r.AddressValidationIdentity(route)
	if identity == nil {
		identity = nil
	}
	return RxDatagram{
		Bytes:                     buffer,
		RouteHandle:               route,
		AddressValidationIdentity: append([]byte(nil), identity...),
	}, nil
}

func (r *UdpRuntime) Wait(nextWakeup coquic.TimeUs, hasNextWakeup bool, idleTimeout time.Duration) (WaitEvent, error) {
	if hasNextWakeup {
		now := r.NowUs()
		if nextWakeup <= now {
			return WaitEvent{Kind: WaitTimer}, nil
		}
		return r.waitWithTimeout(
			time.Duration(uint64(nextWakeup-now))*time.Microsecond,
			WaitTimer,
		)
	}

	return r.waitWithTimeout(idleTimeout, WaitIdle)
}

func (r *UdpRuntime) waitWithTimeout(timeout time.Duration, deadlineKind WaitKind) (WaitEvent, error) {
	if err := r.socket.SetReadDeadline(time.Now().Add(timeout)); err != nil {
		return WaitEvent{}, err
	}
	datagram, err := r.Recv()
	if deadlineErr(err) {
		return WaitEvent{Kind: deadlineKind}, nil
	}
	if err != nil {
		return WaitEvent{}, err
	}
	return WaitEvent{Kind: WaitDatagram, Datagram: datagram}, nil
}

func (r *UdpRuntime) AddressValidationIdentity(routeHandle coquic.RouteHandle) []byte {
	route, ok := r.routesByHandle[routeHandle]
	if !ok {
		return nil
	}
	return route.AddressValidationIdentity
}

func newRuntime(socket *net.UDPConn, recvBufferSize uint64) *UdpRuntime {
	return &UdpRuntime{
		socket:          socket,
		start:           time.Now(),
		routesByHandle:  make(map[coquic.RouteHandle]Route),
		handlesByPeer:   make(map[string]coquic.RouteHandle),
		nextRouteHandle: 1,
		recvBufferSize:  boundedDatagramBufferSize(recvBufferSize),
	}
}

func boundedDatagramBufferSize(size uint64) int {
	if size == 0 || size > MaxUDPDatagramSize {
		return MaxUDPDatagramSize
	}
	return int(size)
}

func resolveRemote(host string, port uint16) (*net.UDPAddr, error) {
	addr, err := net.ResolveUDPAddr("udp", net.JoinHostPort(host, strconv.Itoa(int(port))))
	if err != nil {
		return nil, err
	}
	if addr == nil {
		return nil, fmt.Errorf("failed to resolve remote address")
	}
	return addr, nil
}

func addressValidationIdentity(peer *net.UDPAddr) []byte {
	if ipv4 := peer.IP.To4(); ipv4 != nil {
		identity := make([]byte, 0, 1+4+2)
		identity = append(identity, 0x04)
		identity = append(identity, ipv4...)
		identity = append(identity, byte(peer.Port>>8), byte(peer.Port))
		return identity
	}
	ipv6 := peer.IP.To16()
	identity := make([]byte, 0, 1+16+2)
	identity = append(identity, 0x06)
	identity = append(identity, ipv6...)
	identity = append(identity, byte(peer.Port>>8), byte(peer.Port))
	return identity
}

func deadlineErr(err error) bool {
	if err == nil {
		return false
	}
	netErr, ok := err.(net.Error)
	return ok && netErr.Timeout()
}
